File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Solution {
   
    public String frequencySort(String s) {
        Map<Character , Integer>mp = new HashMap<>();

        for(int i=0;i<s.length();i++){
          mp.put(s.charAt(i), mp.getOrDefault(s.charAt(i),0) + 1);

        }
        ArrayList<int[]> ls = new ArrayList<>();
        for(Map.Entry<Character , Integer> m : mp.entrySet()){
            ls.add(new int[]{m.getKey() - 'a', m.getValue()});
        }
        Collections.sort(ls, (a,b)->b[1] - a[1]);
        StringBuilder ans = new StringBuilder();

        for(int i=0;i<ls.size();i++){
            int n = mp.get((char)('a' + ls.get(i)[0]));

            for(int k=0;k<n;k++){
             ans.append((char)('a' + ls.get(i)[0]));
            }
           
        }
        return ans.toString();

    }
}

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
       vector<int>freq(128,0);
       for(char c:s){
        freq[c]++;
       }      
       auto cmp = [&](char a,char b){
        if(freq[a] == freq[b]) return a<b;
        return freq[a]>freq[b];
       };
       sort(s.begin(),s.end(),cmp);
      return s;
    
    }
};